### PR TITLE
New opc ua client plugin 

### PR DESCRIPTION
**Pull request by fdamador**
related #3041

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by fdamador** [comment]
cannot run `dep ensure -add github.com/gopcua/opcua` because during `dep init` throws error on conflicts with github.com/influxdata/go-syslog/v2 packages having problematic subpackages.

Can someone help?

**Comment by danielnelson** [comment]
Looks like you have it figured out?  For that last failing test you need to add `github.com/gopcua/opcua` to `docs/LICENSE_OF_DEPENDENCIES.md	`

**Comment by fdamador** [comment]
Did you need me to delete the .vscode/launch.json file? I forgot to do that prior to doing a pull request.  Also, at what telegraf version will you release this plugin?

**Comment by jaketimothy** [comment]
Very excited for this.  Will see serious use in factory settings.  Thank you @fdamador and everyone @gopcua

**Comment by MoritzLaut** [comment]
very excited about this too. seems like a lot of users waiting for this plugin!

**Comment by qzheng527** [comment]
Yes, looking forward to the final merge to the next release.

**Comment by haylesnortal** [comment]
Currently seeing the following error

`panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x178b0a9]

goroutine 66 [running]:
github.com/gopcua/opcua/uasc.(*SecureChannel).setState(0x0, 0x2)
	/go/pkg/mod/github.com/gopcua/opcua@v0.1.10/uasc/secure_channel.go:145 +0x9
panic(0x246b040, 0x4556890)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/gopcua/opcua/uasc.(*SecureChannel).hasState(...)
	/go/pkg/mod/github.com/gopcua/opcua@v0.1.10/uasc/secure_channel.go:149
github.com/gopcua/opcua/uasc.(*SecureChannel).closeSecureChannel(0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/gopcua/opcua@v0.1.10/uasc/secure_channel.go:744 +0x9e
github.com/gopcua/opcua/uasc.(*SecureChannel).Close(0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/gopcua/opcua@v0.1.10/uasc/secure_channel.go:665 +0x32
github.com/gopcua/opcua.(*Client).Close(0xc000140a00, 0x1b, 0xc000140b00)
	/go/pkg/mod/github.com/gopcua/opcua@v0.1.10/client.go:234 +0x46
github.com/influxdata/telegraf/plugins/inputs/opcua_client.disconnect(0xc00015c000, 0x2ec0bc0, 0xc00004e6a0)
	/go/src/github.com/influxdata/telegraf/plugins/inputs/opcua_client/opcua_client.go:352 +0xe7
github.com/influxdata/telegraf/plugins/inputs/opcua_client.(*OpcUA).Gather(0xc00015c000, 0x2f370c0, 0xc0003f7600, 0x0, 0x8)
	/go/src/github.com/influxdata/telegraf/plugins/inputs/opcua_client/opcua_client.go:367 +0x37a
github.com/influxdata/telegraf/internal/models.(*RunningInput).Gather(0xc000158320, 0x2f370c0, 0xc0003f7600, 0x2f18020, 0xc0000b4e40)
	/go/src/github.com/influxdata/telegraf/internal/models/running_input.go:115 +0x6d
github.com/influxdata/telegraf/agent.(*Agent).gatherOnce.func1(0xc0000b61e0, 0xc000158320, 0x2f370c0, 0xc0003f7600)
	/go/src/github.com/influxdata/telegraf/agent/agent.go:336 +0x3f
created by github.com/influxdata/telegraf/agent.(*Agent).gatherOnce
	/go/src/github.com/influxdata/telegraf/agent/agent.go:335 +0xf5`

When trying to connect to an opc server.

Using OPC simulator:
`docker run -dt --name plc -p 50000:50000 mcr.microsoft.com/iotedge/opc-plc:1.0.0-linux-amd64 --autoaccept --unsecuretransport`

With Telegraf config:
`
[global_tags]
  dc = "iotedge"

[agent]
  interval = "10s"
  debug = true

[[outputs.file]]
    files = ["stdout"]

[[inputs.opcua_client]]
  name = "plc1"
  endpoint = "opc.tcp://0.0.0.0:50000"
  timeout = 30
  security_policy = "None"
  security_mode = "None"
  nodes = [
      {name="SpikeData", namespace="2", identifier_type="s", identifier="SpikeData", data_type="double", description="Randomly generated data"},
  ]

`

Telegraf code built and pushed to docker hub here:
chrishaylesnortal/telegraf-demo:1.0.1

**Comment by fdamador** [comment]
Chris,

I'm not sure your setup enpoint has a valid IP address. Can you try localhost or 127.0.0.1 or the IP address to the container running your OPC UA server.

> [[inputs.opcua_client]]
> name = "plc1"
**> endpoint = "opc.tcp://0.0.0.0:50000"**

**Comment by haylesnortal** [comment]
@fdamador 

Same result with localhost and 127.0.0.1.

To remove my simulator as a possibility, I tried configuring it to work with a separate KEPWare based simulator, too. Getting the same result there. The Telegraf config I'm using for that:

[global_tags]
  dc = "iotedge"

[agent]
  interval = "10s"
  debug = true

[[outputs.file]]
    files = ["stdout"]

[[inputs.opcua_client]]
  name = "plc1"
  endpoint = "opc.tcp://40.68.7.194:49320"
  timeout = 60
  security_policy = "None"
  security_mode = "None"
  nodes = [
      {name="Tag1Double", namespace="2", identifier_type="s", identifier="SimulationPLC.plugintest.Tag1Double", data_type="double", description=""},
  ]

**Comment by fdamador** [comment]
@haylesnortal

Ok, let me review.  I was able to replicate your original error.  the good news is that I'm able to read the data point using the examples from the gopcua library...so will trace out the bug.

**Comment by haylesnortal** [comment]
@fdamador think I've got it...

1) the error was to do with the disconnect() call before it was ever connected. Commented that out. I was then able to see the real error: regardless of what endpoint address I gave, the client was trying to connect on 127.0.0.1.

2) I checked the endpoints being returned by the OPC UA server and 127.0.0.1 is the first of a few. This is pretty common. And not every server will know it's own public facing endpoint.

3) You were then selecting an endpoint, building out client options from that endpoint (all good up to here) then sending that endpoint.EndpointURL to create the client: `o.client = opcua.NewClient(ep.EndpointURL, opts...)` ...this being the 127.0.0.1 endpoint.

4) So I replace this with the original endpoint supplied by the user: o.client = `opcua.NewClient(o.Endpoint, opts...)`

That seemed to fix it.

**Comment by fdamador** [comment]
Thanks for help, that did it.  I've committed and pushed my changes.

image

**Comment by haylesnortal** [comment]
@fdamador Thanks for taking a look and the quick response. Much appreciated.

**Comment by haylesnortal** [comment]
Still seeing issues... this time, it reads the first value fine, but then it never updates. I can see the OPC sim values changing in my OPC browser, but Telegraf never sees updated values. I think it's because the only call to readvalues() is in the Connect() function. I'll play around with it, see what I can do.

**Comment by fdamador** [comment]
Understood, I made these modifications. Let me know if this is what you were thinking.

File: plugins/inputs/opcua_client/opcua_client.go [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+		return fmt.Errorf("RegisterNodes Read failed: %v", err)
+	}
+	o.ReadSuccess++
+	for i, d := range resp.Results {
+		if d.Status != ua.StatusOK {
+			return fmt.Errorf("Status not OK: %v", d.Status)
+		}
+		o.NodeData[i].TagName = o.NodeList[i].Name
+		if d.Value != nil {
+			o.NodeData[i].Value = d.Value.Value()
```
Comment (russorat) [inline plugins/inputs/opcua_client/opcua_client.go:334]: How does this deal with values like inf (for infinite) or nan (for NaN), or -inf?

Comment (fdamador) [reply plugins/inputs/opcua_client/opcua_client.go:334]: As you can see, it doesn't.  Can you point me to existing plugin code to add?

Comment (russorat) [reply plugins/inputs/opcua_client/opcua_client.go:334]: Looks like the upstream library can definitely return at least NaN: https://github.com/gopcua/opcua/blob/811e88389e4b6e83476a4f33f0e35b44e4d1f9ff/ua/buffer.go#L129

I'm not sure if this plugin needs to worry about it or not TBH. maybe that's the responsibility of a parser or aggregator.

Comment (danielnelson) [reply plugins/inputs/opcua_client/opcua_client.go:334]: Telegraf allows for float fields to be set to NaN and Inf, the serializer for line protocol's current behavior is to log a warning at DEBUG level, and skip the field.

Comment (fdamador) [reply plugins/inputs/opcua_client/opcua_client.go:334]: So, I don’t need to take action on this?

**Comment by PWSys** [comment]
It would be helpful to customize the incoming schema.  Examples include adding custom tags and monitoring multiple fields per measurement.  There won't always be a 1:1 relationship between a monitored OPC-UA node and the desired measuremnt in InfluxDB.

File: plugins/inputs/opcua_client/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#
+#  ## Read Timeout
+#  ## add an arbitrary timeout (seconds) to demonstrate how to stop a subscription
+#  ## with a context.
+timeout = 30
+#
+#  # Time Inteval, default = 100 * time.Millisecond
+#  # interval = "10000000"
+#
+#  # Security policy: None, Basic128Rsa15, Basic256, Basic256Sha256. Default: auto
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/README.md:26]: default says auto, but text says "None" these should match

File: plugins/inputs/opcua_client/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+nodes = [
+		{name="ProductName", namespace="0", identifier_type="i", identifier="2261", data_type="string", description="open62541 OPC UA Server"},
+		{name="ProductUri", namespace="0", identifier_type="i", identifier="2262", data_type="string", description="http://open62541.org"},
+		{name="ManufacturerName", namespace="0", identifier_type="i", identifier="2263", data_type="string", description="open62541"},
+]
+```
+### Example Output:
+
+```
+$ ./telegraf -config telegraf.conf -input-filter opcua_client -test
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/README.md:55]: this should be actual example output, not the command to run to get the output.

File: plugins/inputs/opcua_client/opcua_client_test.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		{"ProductName", "0", "i", "2261", "string", "open62541 OPC UA Server"},
+		{"ProductUri", "0", "i", "2262", "string", "http://open62541.org"},
+		{"ManufacturerName", "0", "i", "2263", "string", "open62541"},
+	}
+
+	var o OpcUA
+	var err error
+
+	o.Name = "testing"
+	o.Endpoint = "opc.tcp://opcua.rocks:4840"
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client_test.go:29]: this is hitting a remote service on the internet? can we store and mock the response instead?

File: plugins/inputs/opcua_client/opcua_client.go [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+	Name           string   `toml:"name"`
+	Endpoint       string   `toml:"endpoint"`
+	SecurityPolicy string   `toml:"security_policy"`
+	SecurityMode   string   `toml:"security_mode"`
+	Certificate    string   `toml:"certificate"`
+	PrivateKey     string   `toml:"private_key"`
+	Interval       string   `toml:"time_interval"`
+	TimeOut        int      `toml:"timeout"`
+	NodeList       []OPCTag `toml:"nodes"`
+	Nodes          []string
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:26]: weird to have NodeList called nodes in TOML, then have an attribute called Nodes. Should all these other attributes be private?

File: plugins/inputs/opcua_client/opcua_client.go [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+	// status
+	ReadSuccess  int
+	ReadError    int
+	NumberOfTags int
+
+	// internal values
+	client *opcua.Client
+	req    *ua.ReadRequest
+	// nodeMonitor       *monitor.NodeMonitor
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:40]: remove

File: plugins/inputs/opcua_client/opcua_client.go [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+#  # OPC UA Endpoint URL
+endpoint = "opc.tcp://opcua.rocks:4840"
+#
+#  ## Read Timeout
+#  ## add an arbitrary timeout (seconds) to demonstrate how to stop a subscription
+#  ## with a context.
+timeout = 30
+#
+#  # Time Inteval, default = 10s
+interval = "1s"
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:98]: should match default

File: plugins/inputs/opcua_client/opcua_client.go [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+		}
+
+		// build nodeid
+		o.Nodes = append(o.Nodes, BuildNodeID(item))
+
+		//parse NodeIds and NodeIds errors
+		nid, niderr := ua.ParseNodeID(o.Nodes[i])
+		// build NodeIds and Errors
+		o.NodeIDs = append(o.NodeIDs, nid)
+		o.NodeIDerror = append(o.NodeIDerror, niderr)
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:247]: should this fail validation?

File: plugins/inputs/opcua_client/opcua_client.go [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+	Certificate    string   `toml:"certificate"`
+	PrivateKey     string   `toml:"private_key"`
+	Interval       string   `toml:"time_interval"`
+	TimeOut        int      `toml:"timeout"`
+	NodeList       []OPCTag `toml:"nodes"`
+	Nodes          []string
+	NodeData       []OPCData
+	NodeIDs        []*ua.NodeID
+	NodeIDerror    []error
+	state          ConnectionState
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:30]: There's no reason to have a connection state, it's just complicating the code. You can use `o.client == nil` to determine if you need to try to connect.

File: plugins/inputs/opcua_client/opcua_client.go [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+			return fmt.Errorf("RegisterNodes failed: %v", err)
+		}
+
+		o.req = &ua.ReadRequest{
+			MaxAge:             2000,
+			NodesToRead:        readvalues(regResp.RegisteredNodeIDs),
+			TimestampsToReturn: ua.TimestampsToReturnBoth,
+		}
+
+		err = o.getData()
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:310]: Connect should not call getData, gather already does this.

File: plugins/inputs/opcua_client/opcua_client.go [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+	return rvids
+}
+
+func disconnect(o *OpcUA) error {
+	u, err := url.Parse(o.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	o.ReadError = 0
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:358]: I don't think there's any reason to reset these.

File: plugins/inputs/opcua_client/opcua_client.go [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+	o.ReadSuccess = 0
+
+	switch u.Scheme {
+	case "opc.tcp":
+		o.state = Disconnected
+		o.client.Close()
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:364]: unset o.client here so you know you need to reconnect. and you can drop o.state.

File: plugins/inputs/opcua_client/opcua_client.go [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+func readvalues(ids []*ua.NodeID) []*ua.ReadValueID {
+	rvids := make([]*ua.ReadValueID, len(ids))
+	for i, v := range ids {
+		rvids[i] = &ua.ReadValueID{NodeID: v}
+	}
+	return rvids
+}
+
+func disconnect(o *OpcUA) error {
+	u, err := url.Parse(o.Endpoint)
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:353]: There's no reason to parse the endpoint or check the scheme here.

File: plugins/inputs/opcua_client/opcua_client.go [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+			return fmt.Errorf("Status not OK: %v", d.Status)
+		}
+		o.NodeData[i].TagName = o.NodeList[i].Name
+		if d.Value != nil {
+			o.NodeData[i].Value = d.Value.Value()
+			o.NodeData[i].DataType = d.Value.Type()
+		}
+		o.NodeData[i].Quality = d.Status
+		o.NodeData[i].TimeStamp = d.ServerTimestamp.String()
+		o.NodeData[i].Time = d.SourceTimestamp.String()
```
Comment (ssoroka) [inline plugins/inputs/opcua_client/opcua_client.go:339]: string timestamps are notoriously difficult to work with. Not sure why you'd convert to a string here.

File: plugins/inputs/opcua_client/README.md (hunk 3 above)
Comment (sjwang90) [reply plugins/inputs/opcua_client/README.md:55]: an example input/output would be very useful for this README

File: plugins/inputs/opcua_client/README.md [hunk 15]
```diff
@@ -0,0 +1,57 @@
+# Telegraf Input Plugin: opcua_client
+
+The opcua_client plugin retrieves data from OPCUA slave devices
```
Comment (sjwang90) [inline plugins/inputs/opcua_client/README.md:3]: Please replace the term `slave` with `replica`, `secondary` or any of the following [suggestions](https://tools.ietf.org/id/draft-knodel-terminology-00.html#suggested-alternatives).

Comment (fdamador) [reply plugins/inputs/opcua_client/README.md:3]: Thanks I’ll replace with server to match industry standards nomenclature

**Comment by sjwang90** [comment]
Hey @fdamador, just checking in on the progress of @ssoroka's suggestions. We'd love to get others to start using this! Thanks again!

**Comment by sjwang90** [comment]
@fdamador do you think you could test your plugin with the [Prosys simulator](https://prosysopc.com/products/opc-ua-simulation-server/)? Was trying to troubleshoot with @haylesnortal last week and was getting various errors.

`2020-07-21T22:16:05Z E! [inputs.opcua_client] Error in plugin: Endpoint Error: opcua: could not resolve address ip-192-168-1-2.ec2.internal:53530`

**Comment by sjwang90** [comment]
Hey @fdamador, we merged in #8009 last week. Wanted to give you a big thanks as your plugin was the base of the final plugin that was merged. 

Feel free to test out that new plugin with master or the nightly build.

**Comment by fdamador** [comment]
Awesome!  Glad I could be of service.

**Pull request closed without merging**